core/chains/evm/client/rpc_client.go

```diff
@@ -0,0 +1,945 @@
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+
+	clienttypes "github.com/smartcontractkit/chainlink/v2/common/chains/client"
+	"github.com/smartcontractkit/chainlink/v2/core/assets"
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+type RPCClient interface {
+	clienttypes.NodeClientAPI[
+		*big.Int,
+		common.Hash,
+		*evmtypes.Head,
+		ethereum.Subscription,
+	]
+	clienttypes.RPCClient[
+		*big.Int,
+		evmtypes.Nonce,
+		common.Address,
+		common.Hash,
+		types.Transaction,
+		common.Hash,
+		types.Log,
+		ethereum.FilterQuery,
+		*evmtypes.Receipt,
+		*assets.Wei,
+		*evmtypes.Head,
+		ethereum.Subscription,
+	]
+}
+
+type rpcClient struct {
+	utils.StartStopOnce
+	lfcLog  logger.Logger
+	rpcLog  logger.Logger
+	name    string
+	id      int32
+	chainID *big.Int
+
+	ws   rawclient
+	http *rawclient
+
+	stateMu sync.RWMutex // protects state* fields
+	state   clienttypes.NodeState
+
+	// Need to track subscriptions because closing the RPC does not (always?)
+	// close the underlying subscription
+	subs []ethereum.Subscription
+
+	// chStopInFlight can be closed to immediately cancel all in-flight requests on
+	// this rpcClient. Closing and replacing should be serialized through
+	// stateMu since it can happen on state transitions as well as rpcClient Close.
+	chStopInFlight chan struct{}
+	// rpcClientCtx is the rpcClient lifetime's context
+	rpcClientCtx context.Context
+	// cancelRpcClientCtx cancels rpcClientCtx when stopping the rpcClient
+	cancelRpcClientCtx context.CancelFunc
+	// wg waits for subsidiary goroutines
+	wg sync.WaitGroup
+}
+
+// NewRPCCLient returns a new *rpcClient as clienttypes.RPCClient
+func NewRPCClient(lggr logger.Logger, wsuri url.URL, httpuri *url.URL, name string, id int32, chainID *big.Int) RPCClient {
+	r := new(rpcClient)
+	r.name = name
+	r.id = id
+	r.chainID = chainID
+	r.ws.uri = wsuri
+	if httpuri != nil {
+		r.http = &rawclient{uri: *httpuri}
+	}
+	r.chStopInFlight = make(chan struct{})
+	r.rpcClientCtx, r.cancelRpcClientCtx = context.WithCancel(context.Background())
+	lggr = lggr.Named("Client").With(
+		"clientTier", "primary",
+		"clientName", name,
+		"client", r.String(),
+		"evmChainID", chainID,
+	)
+	r.lfcLog = lggr.Named("Lifecycle")
+	r.rpcLog = lggr.Named("RPC")
+
+	return r
+}
+
+// Not thread-safe
+// Pure dial: does not mutate rpcClient "state" field.
+func (r *rpcClient) Dial(callerCtx context.Context) error {
+	ctx, cancel := r.makeQueryCtx(callerCtx)
+	defer cancel()
+
+	promEVMPoolRPCNodeDials.WithLabelValues(r.chainID.String(), r.name).Inc()
+	lggr := r.lfcLog.With("wsuri", r.ws.uri.Redacted())
```

Comment:
Remove lfcLog and use rpcLog instead. Lifecycle logs are handled by the caller. At this level, everything is considered an RPC call.